File: project/backend-services/smart-irrigation-backend/domain-services/src/main/java/pt/sensae/services/smart/irrigation/backend/domainservices/garden/mapper/GardenMapper.java
package pt.sensae.services.smart.irrigation.backend.domainservices.garden.mapper;

import pt.sensae.services.smart.irrigation.backend.domain.model.GPSPoint;
import pt.sensae.services.smart.irrigation.backend.domain.model.business.garden.*;
import pt.sensae.services.smart.irrigation.backend.domainservices.garden.model.BoundaryCommandDetails;
import pt.sensae.services.smart.irrigation.backend.domainservices.garden.model.CreateGardeningAreaCommand;
import pt.sensae.services.smart.irrigation.backend.domainservices.garden.model.UpdateGardeningAreaCommand;

import java.util.List;
import java.util.stream.Collectors;

public class GardenMapper {

    public static GardeningArea toModel(UpdateGardeningAreaCommand updateCommand) {
        var of = GardeningAreaId.of(updateCommand.id);
        return getGardeningArea(updateCommand.area, of, updateCommand.name);
    }

    public static GardeningArea toModel(CreateGardeningAreaCommand createCommand) {
        var of = GardeningAreaId.create();
        return getGardeningArea(createCommand.area, of, createCommand.name);
    }

    private static GardeningArea getGardeningArea(List<BoundaryCommandDetails> boundaryCommands, GardeningAreaId of, String name) {
        var area = new Area(boundaryCommands.stream().map(b -> {
            var gps = new GPSPoint(b.latitude, b.longitude, b.altitude);
            return new BoundaryPoint(b.position, gps);
        }).collect(Collectors.toSet()));

        return new GardeningArea(of, GardenName.of(name), area, Owners.empty());
    }
}

File: project/frontend-services/libs/smart-irrigation/mapper/src/lib/mappers/GardenMapper.ts
import {GardeningAreaDTO} from "@frontend-services/smart-irrigation/dto";
import {
  AreaBoundary,
  GardeningArea,
  GardeningAreaId,
  GardeningAreaName,
  GPSDataDetails
} from "@frontend-services/smart-irrigation/model";

export class GardenMapper {
  static dtoToModel(dto: GardeningAreaDTO): GardeningArea {
    const area = dto.area.map(b => new AreaBoundary(b.position, new GPSDataDetails(b.latitude, b.longitude, b.altitude)));
    return new GardeningArea(new GardeningAreaId(dto.id), new GardeningAreaName(dto.name), area);
  }
}
